### PR TITLE
Add `withClamp` documentation

### PR DESCRIPTION
**Pull request by Latropos**
## Summary
Add documentation of withClamp animation modifier

### Interactive playround
|☀️|🌙|
|--|--|
|image|image|

### CODE REVIEW

File: docs/docs/animations/withClamp.mdx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+| Name             | Type Description |
+| ---------------- | ---------------- | ------------------------------------------------ |
+| min <Optional /> | `number`         | The lowest value your animation can ever reach   |
+| max <Optional /> | `number`         | The greatest value your animation can ever reach |
+
+#### `animation`
+
+The spring animation you want to clamp.
+
```
Comment (kacperkapusciak) [inline docs/docs/animations/withClamp.mdx:28]: A real-life example of use that's copy-paste-runable is more than welcome here

File: docs/src/components/InteractivePlayground/useClampPlayground/index.jsx
[the remark this answers is not in the document]
Comment (kacperkapusciak) [reply docs/src/components/InteractivePlayground/useClampPlayground/index.jsx:1]: And could we have that written in TypeScript? 🙏

File: docs/src/components/InteractivePlayground/useClampPlayground/index.jsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    setUpperSpringToValue(initialState.upperSpringToValue);
+  };
+
+  const code = `width: withClamp({
+    min: ${lowerBound}
+    max: ${upperBound}
+  },
+  withSpring(width.value, config)
+)
+`
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/index.jsx:59]:
```suggestion
  const code = `{
  width: withClamp(
    { min: ${lowerBound}, max: ${upperBound} },
    withSpring(width.value)
  ),
}
`
```

A little bit more readable

File: docs/docs/animations/withClamp.mdx (hunk 1 above)
Comment (kacperkapusciak) [inline docs/docs/animations/withClamp.mdx:28]: and it's missing what this function is returning

File: docs/src/components/InteractivePlayground/useClampPlayground/index.jsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    <DoubleRange
+      label="Clamp limits"
+      min={0}
+      max={400}
+      step={10}
+      value={[lowerBound, upperBound]}
+      onChange={[setLowerBound, setUpperBound]}
+    />
+    <DoubleRange
+      label="Scope of movement"
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/index.jsx:36]: I don't like this label as the scope of movement says nothing. This is the value to which the spring is finishing at 🤔

File: docs/src/components/InteractivePlayground/useClampPlayground/example.tsx
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/example.tsx:1]: Files with components usually start with a capital letter

File: docs/src/components/InteractivePlayground/useClampPlayground/example.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  lowerBound: number;
+  upperBound: number;
+  lowerSpringToValue: number;
+  upperSpringToValue: number;
+}
+
+interface Props {
+  options: ClampPlaygroundOptions;
+}
+export default function useClampPlayground({ options }): Props {
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/example.tsx:25]:
```suggestion
export default function App({ options }): Props {
```

File: docs/src/components/InteractivePlayground/useClampPlayground/example.tsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        }}
+      />
+    </View>
+  );
+}
+
+const styles = StyleSheet.create({
+  container: {
+    flex: 1,
+    flexDirection: 'column',
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/example.tsx:129]:
```suggestion
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
```

lets center it a bit

File: docs/src/components/InteractivePlayground/useClampPlayground/example.tsx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    );
+  }
+
+  return (
+    <View style={styles.container}>
+      {renderExample(clampedStyle, 'Clamped spring', true)}
+      {renderExample(defaultStyle, 'Default spring', false)}
+
+      <Button
+        title="toggle"
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/example.tsx:114]: please add a vertical margin to this button 🙏

**Review by kacperkapusciak — CHANGES_REQUESTED** [review]

File: docs/src/components/InteractivePlayground/index.tsx [hunk 7]
```diff
@@ -162,6 +166,56 @@ export function Range({
...
+        <label>{label}</label>
+        {[0, 1].map((idx) => {
+          return (
+            <TextField
+              type="number"
+              hiddenLabel
+              size="small"
+              inputProps={{ min: min, max: max, step: step }}
+              sx={TextFieldStyling}
+              value={value[idx]}
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/index.tsx:189]: You're also missing a `key` prop on a loop

File: docs/docs/animations/withClamp.mdx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+```typescript
+type AnimatableValue = number | string | number[];
+
+function withClamp= <T extends number | string>(
+    config: {
+        min?: T;
+        max?: T;
+    },
+    clampedAnimation: T): T;
```
Comment (kacperkapusciak) [inline docs/docs/animations/withClamp.mdx:36]: formatting and typo
```suggestion
function withClamp <T extends number | string>(config: {
    min?: T;
    max?: T;
}, clampedAnimation: T) => T;
```

**Review by kacperkapusciak — CHANGES_REQUESTED** [review]

**Comment by piaskowyk** [comment]
I think that this example is too complex. Although I understand how clamp works, but when I look at this particular example, I feel confused. Is it possible to simplify it? Maybe we can use vertical lines as a clamp boundaries instead of rectangles, similar to the example with interpolation. cc @kacperkapusciak

**Comment by Latropos** [comment]
@piaskowyk
Is it less confusing with color hints?
image

File: docs/docs/animations/withClamp.mdx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+<details>
+<summary>Type definitions</summary>
+
+```typescript
+type AnimatableValue = number | string | number[];
+
+function withClamp= <T extends number | string>( config: {
+        min?: T;
+        max?: T;
+    }, clampedAnimation: T): T;
```
Comment (kacperkapusciak) [inline docs/docs/animations/withClamp.mdx:34]:
```suggestion
function withClamp<T extends number | string>(config: {
  min?: T;
  max?: T;
}, clampedAnimation: T): T;
```

File: docs/src/components/InteractivePlayground/useClampPlayground/Example.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          withSequence(
+            withDelay(2000, withSpring(BOX_STOP, config)),
+            withTiming(BOX_START, { duration: 0 })
+          ),
+          -1,
+          true
+        )
+      ),
+    };
+  });
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/Example.tsx:50]:
```suggestion
  const clampedStyle = useAnimatedStyle(() => {
    return {
      transform: [
        {
          translateX: withClamp(
            { min: options.lowerBound, max: options.upperBound },
            withRepeat(
              withSequence(
                withDelay(2000, withSpring(BOX_STOP, config)),
                withTiming(BOX_START, { duration: 0 })
              ),
              -1,
              true
            )
          ),
        },
      ],
    };
  });
```

Always prefer `transform` to position props like `left`, `right`, `top`, `bottom` as they trigger layouting process and can cause jitter (see the borders)

`left`:

https://github.com/software-mansion/react-native-reanimated/assets/39658211/815dddf8-37e8-4eb7-ab1f-26dc214eda43


`translateX`:

https://github.com/software-mansion/react-native-reanimated/assets/39658211/a8f6ffba-9f2f-4167-bc41-7ab1bb3222d8

File: docs/src/components/InteractivePlayground/useClampPlayground/Example.tsx [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      left: withRepeat(
+        withSequence(
+          withDelay(2000, withSpring(200, config)),
+          withTiming(0, { duration: 0 })
+        ),
+        -1,
+        true
+      ),
+    };
+  });
```
Comment (kacperkapusciak) [inline docs/src/components/InteractivePlayground/useClampPlayground/Example.tsx:62]:
```suggestion
  const defaultStyle = useAnimatedStyle(() => {
    return {
      transform: [
        {
          translateX: withRepeat(
            withSequence(
              withDelay(2000, withSpring(200, config)),
              withTiming(0, { duration: 0 })
            ),
            -1,
            true
          ),
        },
      ],
    };
  });
```

**Review by kacperkapusciak — APPROVED** [review]
🎉

**Pull request merged**